pkg/cstorpool/v1alpha2/kubernetes_test.go

```diff
@@ -0,0 +1,422 @@
+}
+
+func fakeGetClientsetNil() (clientset *clientset.Clientset, err error) {
+	return nil, nil
+}
+
+func fakeGetClientsetErr() (clientset *clientset.Clientset, err error) {
+	return nil, errors.New("Some error")
+}
+
+func TestWithDefaults(t *testing.T) {
+	tests := map[string]struct {
+		listFn             listFunc
+		getFn              getFunc
+		getClientsetFn     getClientsetFunc
+		createFn           createFunc
+		patchFn            patchFunc
+		deleteFn           delFn
+		expectList         bool
+		expectGet          bool
+		expectGetClientset bool
+		expectCreate       bool
+		expectPatch        bool
+		expectDelete       bool
+	}{
+		"When mockclient is empty": {nil, nil, nil, nil, nil, nil, true, true, true, true, true, true},
+		"When mockclient contains all of them": {fakeListOk, fakeGetOk,
+			fakeGetClientsetOk, fakeCreateOk, fakePatchOk, fakeDeleteOk, true, true, true, true, true, true},
+	}
+
+	for name, mock := range tests {
+		name := name //pin it
+		mock := mock // pin it
+		t.Run(name, func(t *testing.T) {
+			fc := &Kubeclient{}
+			fc.list = mock.listFn
+			fc.get = mock.getFn
+			fc.getClientset = mock.getClientsetFn
+			fc.create = mock.createFn
+			fc.patch = mock.patchFn
+
+			fc.withDefaults()
+
+			if (fc.list != nil) != mock.expectList {
+				t.Fatalf("test %s failed: expected non-nil fc.list but got %v", name, fc.list)
+			}
+
+			if (fc.get != nil) != mock.expectGet {
+				t.Fatalf("test %s failed: expected non-nil fc.get but got %v", name, fc.get)
+			}
+
+			if (fc.getClientset != nil) != mock.expectGetClientset {
+				t.Fatalf("test %s failed: expected non-nil fc.getClientset but got %v", name, fc.getClientset)
+			}
+
+			if (fc.create != nil) != mock.expectCreate {
+				t.Fatalf("test %s failed: expected non-nil fc.create but got %v", name, fc.create)
+			}
+
+			if (fc.patch != nil) != mock.expectPatch {
+				t.Fatalf("test %s failed: expected non-nil fc.patch but got %v", name, fc.patch)
+			}
+
+			if (fc.del != nil) != mock.expectDelete {
+				t.Fatalf("test %s failed: expected non-nil fc.patch but got %v", name, fc.del)
+			}
+		})
+	}
+}
+
+func TestWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		clientSet    *clientset.Clientset
+		isKubeClient bool
+	}{
+		"Clientset is empty":     {nil, false},
+		"Clientset is not empty": {&clientset.Clientset{}, true},
+	}
+
+	for name, mock := range tests {
+		name := name //pin it
+		mock := mock // pin it
+		t.Run(name, func(t *testing.T) {
+			h := WithClientset(mock.clientSet)
+			fake := &Kubeclient{}
+			h(fake)
+
+			if mock.isKubeClient && fake.clientset == nil {
+				t.Fatalf("test %s failed, expected non-nil fake.clientset but got %v", name, fake.clientset)
+			}
+
+			if !mock.isKubeClient && fake.clientset != nil {
+				t.Fatalf("test %s failed, expected nil fake.clientset but got %v", name, fake.clientset)
+			}
+		})
+	}
+}
+
+func TestGetClientOrCached(t *testing.T) {
+	tests := map[string]struct {
+		kubeClient *Kubeclient
```

Comment:
nit:
IMO it is good to have the only the properties which are useful to get clientset
```
tests := map[string]struct{
             clientset *clientset.Clientset
             getClientset getClientsetFunc
            }{
             // mock over these two is good enough
            }
So In feature if any property of kubeclient is add no need to update the table.

Reply:
will do in next PR.